crates/ra_vfs/src/io.rs

```diff
@@ -46,7 +47,40 @@ enum ChangeKind {
 
 const WATCHER_DELAY: Duration = Duration::from_millis(250);
 
-pub(crate) type Worker = thread_worker::Worker<Task, TaskResult>;
+// Like thread::JoinHandle, but joins the thread on drop.
+//
+// This is useful because it guarantees the absence of run-away threads, even if
+// code panics. This is important, because we might seem panics in the test and
+// we might be used in an IDE context, where a failed component is just
+// restarted.
+//
+// Because all threads are joined, care must be taken to avoid deadlocks. That
+// typically means ensuring that channels are dropped before the threads.
+struct ScopedThread(Option<thread::JoinHandle<()>>);
+
+impl ScopedThread {
+    fn spawn(name: String, f: impl FnOnce() + Send + 'static) -> ScopedThread {
+        let handle = thread::Builder::new().name(name).spawn(f).unwrap();
+        ScopedThread(Some(handle))
+    }
+}
+
+impl Drop for ScopedThread {
+    fn drop(&mut self) {
+        let res = self.0.take().unwrap().join();
+        if !thread::panicking() {
+            res.unwrap();
+        }
+    }
+}
+
+pub(crate) struct Worker {
+    // XXX: it's important to drop `sender` before `_thread` to avoid deadlock.
```

Comment:
you might consider including the detail that the desired drop order is accomplished here via the field declaration order. At least, I don't know if its something that is considered common knowledge in the Rust dev community. (The original `Worker` docs [here](https://github.com/rust-analyzer/rust-analyzer/blob/bd2b4ef2d1c631215523f79a8138cfa8cbd3f70e/crates/thread_worker/src/lib.rs#L42) did include that detail.)